src/Tools/ExternalAccess/LiveShare/RemoteLanguageServiceWorkspace.cs

```diff
@@ -0,0 +1,444 @@
+            if (string.IsNullOrEmpty(text.ToString()))
+            {
+                text = SourceText.From(File.ReadAllText(document.FilePath));
+
+                // Some features like the FindRefs window try to get the text at the span without opening the document (for eg to classify the span).
+                // So fork the document to get one with the text. Note that this new document will not be in the CurrentSolution and we don't intend to
+                // apply it back. By fetching the file, the workspace will get updated anyway. The assumption here is that this document that we are
+                // handing out is only used for simple inspection and it's version is never compared with the Workspace.CurrentSolution.
+                document = document.WithText(text);
+            }
+
+            var textSpan = ProtocolConversions.RangeToTextSpan(location.Range, text);
+            return new DocumentSpan(document, textSpan);
+        }
+
+        private Document AddDocumentToProject(string filePath, string language, string projectName)
+        {
+            var project = CurrentSolution.Projects.FirstOrDefault(p => p.Name == projectName && p.Language == language);
+            if (project == null)
+            {
+                var projectInfo = ProjectInfo.Create(ProjectId.CreateNewId(), VersionStamp.Create(), projectName, projectName, language);
+                OnManagedProjectAdded(projectInfo);
+                project = CurrentSolution.GetProject(projectInfo.Id);
+            }
+
+            var docInfo = DocumentInfo.Create(DocumentId.CreateNewId(project.Id),
+                                                  name: Path.GetFileName(filePath),
+                                                  loader: new FileTextLoader(filePath, null),
+                                                  filePath: filePath);
+            OnDocumentAdded(docInfo);
+            return CurrentSolution.GetDocument(docInfo.Id);
+        }
+
+        private string GetLanguage(string filePath)
+        {
+            var fileExtension = Path.GetExtension(filePath).ToLower();
+
+            if (fileExtension == ".cs")
+            {
+                return StringConstants.CSharpLspLanguageName;
+            }
+            else if (fileExtension == ".ts" || fileExtension == ".js")
+            {
+                return StringConstants.TypeScriptLanguageName;
+            }
+
+            return null;
+        }
+
+        /// <inheritdoc />
+        public void NotifyOnDocumentClosing(string moniker)
+        {
+            if (_openedDocs.TryGetValue(moniker, out DocumentId id))
+            {
+                // check if the doc is part of the current Roslyn workspace before notifying Roslyn.
+                if (CurrentSolution.ContainsProject(id.ProjectId))
+                {
+                    OnDocumentClosed(id, new FileTextLoaderNoException(moniker, null));
+                    _openedDocs = _openedDocs.Remove(moniker);
```

Comment:
when do you modify _openedDocs directly and when you use RemoveOpenedDocument stuff?